sound/soc/intel/boards/cml_rt1011_rt5682.c

```diff
@@ -0,0 +1,477 @@
+	ret = snd_soc_component_set_jack(component, jack, NULL);
+	if (ret)
+		dev_err(rtd->dev, "Headset Jack call-back failed: %d\n", ret);
+
+	return ret;
+};
+
+static int cml_rt5682_hw_params(struct snd_pcm_substream *substream,
+				struct snd_pcm_hw_params *params)
+{
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct snd_soc_dai *codec_dai = rtd->codec_dai;
+	int clk_id, clk_freq, pll_out, ret;
+
+	clk_id = RT5682_PLL1_S_MCLK;
+	clk_freq = CML_PLAT_CLK;
+
+	pll_out = params_rate(params) * 512;
+
+	ret = snd_soc_dai_set_pll(codec_dai, 0, clk_id, clk_freq, pll_out);
+	if (ret < 0)
+		dev_err(rtd->dev, "snd_soc_dai_set_pll err = %d\n", ret);
+
+	/* Configure sysclk for codec */
+	ret = snd_soc_dai_set_sysclk(codec_dai, RT5682_SCLK_S_PLL1,
+				     pll_out, SND_SOC_CLOCK_IN);
+	if (ret < 0)
+		dev_err(rtd->dev, "snd_soc_dai_set_sysclk err = %d\n", ret);
+
+	/*
+	 * slot_width should be equal or large than data length, set them
+	 * be the same
+	 */
+	ret = snd_soc_dai_set_tdm_slot(codec_dai, 0x0, 0x0, 2,
+				       params_width(params));
+	if (ret < 0) {
+		dev_err(rtd->dev, "set TDM slot err:%d\n", ret);
+	}
+	return ret;
+}
+
+static int cml_rt1011_hw_params(struct snd_pcm_substream *substream,
+				struct snd_pcm_hw_params *params)
+{
+	struct snd_soc_pcm_runtime *rtd = substream->private_data;
+	struct snd_soc_dai *codec_dai = rtd->codec_dai;
+	struct snd_soc_card *card = rtd->card;
+	int srate, i, ret;
+
+	srate = params_rate(params);
+
+	for (i = 0; i < rtd->num_codecs; i++) {
+		codec_dai = rtd->codec_dais[i];
+
+		/* 100 Fs to drive 24 bit data */
+		ret = snd_soc_dai_set_pll(codec_dai, 0, RT1011_PLL1_S_BCLK,
+					  100 * srate, 256 * srate);
+		if (ret < 0) {
+			dev_err(card->dev, "codec_dai clock not set\n");
+			return ret;
+		}
+
+		ret = snd_soc_dai_set_sysclk(codec_dai,
+					     RT1011_FS_SYS_PRE_S_PLL1,
+					     256 * srate, SND_SOC_CLOCK_IN);
+		if (ret < 0) {
+			dev_err(card->dev, "codec_dai clock not set\n");
+			return ret;
+		}
+
+		/* TDM 4 slots 24 bit */
+		if (!strcmp(codec_dai->component->name, "i2c-10EC1011:00")) {
+			ret = snd_soc_dai_set_tdm_slot(codec_dai,
+						       0x4, 0x1, 4, 24);
+			if (ret < 0) {
+				dev_err(rtd->dev,
+					"can't set codec TDM slot %d\n", ret);
+				return ret;
+			}
+		}
+		if (!strcmp(codec_dai->component->name, "i2c-10EC1011:01")) {
+			ret = snd_soc_dai_set_tdm_slot(codec_dai,
+						       0x8, 0x2, 4, 24);
+			if (ret < 0) {
+				dev_err(rtd->dev,
+					"can't set codec TDM slot %d\n", ret);
+				return ret;
+			}
+		}
+		if (!strcmp(codec_dai->component->name, "i2c-10EC1011:02")) {
+			ret = snd_soc_dai_set_tdm_slot(codec_dai,
+						       0x1, 0x1, 4, 24);
+			if (ret < 0) {
+				dev_err(rtd->dev,
+					"can't set codec TDM slot %d\n", ret);
+				return ret;
+			}
+		}
+		if (!strcmp(codec_dai->component->name, "i2c-10EC1011:03")) {
+			ret = snd_soc_dai_set_tdm_slot(codec_dai,
+						       0x2, 0x2, 4, 24);
```

Comment:
that RX slot is the same as the :01 device, so this cannot possibly work.
Why can't you use a more regular loop and slot assignments instead of those comparisons on the name and different slots for RX and TX?

Reply:
The RX slot for 2 codecs will be same as its a 2 CH playback over 4 codecs. But each codec provides its own feedback. So each codec Tx is unique, but 2 codecs will have same Rx.

Reply:
Just to add more details, the 2 CH playback is done over 4 codecs - 2 Woofers and 2 Tweeters. 
The Left woofer and tweeter plays the Left playback data and similar by the Right. The RT1011 internally manages control of the Tweeter & Woofer and parameters for the same.

Hence 2 codecs (1 T and 1 W pair) share same Rx slot. I've updated the title too explaining we  are playing 2 CH TDM. Feedback is 4 CH.

Let me know if any more details needed.

Reply:
in that case, please make it explicit which codecs use what slot for RX, e.g. left RX first and then right RX second. If reviewers can't figure out a feature and think you've done a mistake there's something missing.

Reply:
Sure Pierre. I've updated with explaining the rationale and grouping based on Rx slot as suggested.
I